pkg/chunked/storage_linux.go

```diff
@@ -146,40 +146,48 @@ func (c *chunkedDiffer) convertTarToZstdChunked(destDirectory string, payload *o
 func GetDiffer(ctx context.Context, store storage.Store, blobDigest digest.Digest, blobSize int64, annotations map[string]string, iss ImageSourceSeekable) (graphdriver.Differ, error) {
 	pullOptions := store.PullOptions()
 
+	convertImages := parseBooleanPullOption(pullOptions, "convert_images", false)
+
 	if !parseBooleanPullOption(pullOptions, "enable_partial_images", true) {
-		return nil, errors.New("enable_partial_images not configured")
+		return nil, internal.ErrPartialContentNotAvailable
 	}
 
 	zstdChunkedTOCDigestString, hasZstdChunkedTOC := annotations[internal.ManifestChecksumKey]
 	estargzTOCDigestString, hasEstargzTOC := annotations[estargz.TOCJSONDigestAnnotation]
 
 	if hasZstdChunkedTOC && hasEstargzTOC {
-		return nil, errors.New("both zstd:chunked and eStargz TOC found")
+		if !convertImages {
+			return nil, fmt.Errorf("both zstd:chunked and eStargz TOC found: %w", internal.ErrPartialContentNotAvailable)
+		}
+		// If a conversion is required, ignore both TOCs and retrieve the entire layer.
+		hasZstdChunkedTOC, hasEstargzTOC = false, false
 	}
 
 	if hasZstdChunkedTOC {
 		zstdChunkedTOCDigest, err := digest.Parse(zstdChunkedTOCDigestString)
-		if err != nil {
-			return nil, fmt.Errorf("parsing zstd:chunked TOC digest %q: %w", zstdChunkedTOCDigestString, err)
+		if err == nil {
```

Comment:
(I weakly think that if the TOC digest annotations exist but are invalid, we should fail without fallback. If nothing else, aesthetically, there are quite a few `if !convertImages` conditions here.)

Reply:
IMO when partial pulls are not enforced, the entire feature is just an optimization and I think we should be as forgiving as possible.

In any case, we can relax it later if needed and return a hard failure for now

Reply:
I don’t feel strongly about this. “First, do no harm” makes sense as a priority.

I’m thinking that if something is legitimately producing invalid digests, that’s _never_ going to trigger partial pulls and the authors of that image writer should be told, so that they fix things. But, to be fair, that is a notification of developers via users seeing failures, and users might have a good reason to not like that.